### PR TITLE
Capture output

### PR DESCRIPTION
**Pull request by 00vareladavid**
Aiming for #122 with this one.

Basically just add an `io` field to `Context`, thread the `Context` object through a more functions, and make the display stuff use the `io` field by default. I also switched some functions from taking `EnvCache` to `Context`. Most were necessary because they eventually printed, but some I just changed for consistency's sake.

This should also let us test `Display` more easily, which we are not really doing right now.

### CODE REVIEW

**Comment by fredrikekre** [comment]
It would be good to decide if we are gonna use `stdout` or the logging system also, right now it is very random what information is sent where.

**Comment by KristofferC** [comment]
FWIW, I think we should just ditch the logging completely.

**Comment by StefanKarpinski** [comment]
I think it would be good to silently log but print info separately.

**Comment by tkoolen** [comment]
Just wanted to bump this PR. I often need Pkg operations to be completely silent, which is currently hard to do given the lack of an easy way to suppress output in general, see https://github.com/JuliaLang/julia/issues/22879 and the fact that Suppressor.jl exists and is not trivial.

FWIW, I like the logging approach precisely because it's easy to control verbosity: no need to learn a new API, no way to forget to pass a `Context`, and it just seems like it's exactly what the logging system is meant for; a good non-`Logging`-based solution is bound to re-implement a lot of `Logging`'s features. But mostly I just care about there being *a* solution.

**Comment by fredrikekre** [comment]
```
$ jlpkg add Example > /dev/null
```
:)

**Comment by 00vareladavid** [comment]
bors try

**Comment by 00vareladavid** [comment]
bors try

**Comment by 00vareladavid** [comment]
bors try

**Comment by 00vareladavid** [comment]
bors try

**Comment by 00vareladavid** [comment]
bors try

**Comment by 00vareladavid** [comment]
bors try

File: src/Types.jl [hunk 1]
```diff
@@ -309,6 +309,7 @@ end
 # ENV variables to set some of these defaults?
 Base.@kwdef mutable struct Context
     env::EnvCache = EnvCache()
+    io::IO = stdout
```
Comment (KristofferC) [inline src/Types.jl]: I think we should change this to `stderr`.

Comment (00vareladavid) [reply src/Types.jl]: Please explain. Its not weird to output to `stderr` by default? What is the benefit over `stdout`?

Comment (KristofferC) [reply src/Types.jl]: I don't think it is that weird. Many unix applications work by piping their stdout to each other. If an application starts spewing loginfo to stdout then that breaks. The loggers in Base already use stderr so I don't think it is weird for us to also log to that.

Comment (fredrikekre) [reply src/Types.jl]: But e.g. for `Pkg.status` it would make sense with `stdout`, no?

Comment (KristofferC) [reply src/Types.jl]: Tbh, I'm not sure.

**Comment by 00vareladavid** [comment]
Anyone mind if I merge? My main concern here is getting the `io` field to all the print statements. We can split the output into user feedback vs actual logging at a later time.

**Comment by KristofferC** [comment]
Go for it

File: src/Types.jl (hunk 1 above)
Comment (GunnarFarneback) [reply src/Types.jl]: What would make sense for `Pkg.status` to do would be to print nothing and return some kind of object with the status information. Then the REPL could have it printed the usual way with the `show` method, the Pkg REPL mode could print it any way it finds useful, and a program could look into the object and draw whatever conclusions it finds meaningful.

**Pull request merged**